typescript/agentkit/src/wallet-providers/privyEvmEmbeddedWalletProvider.ts

```diff
@@ -0,0 +1,517 @@
+   * @returns The hash of the transaction.
+   */
+  async sendTransaction(transaction: TransactionRequest): Promise<Hex> {
+    const body = {
+      address: this.#address,
+      chain_type: "ethereum",
+      method: "eth_sendTransaction",
+      params: {
+        transaction: {
+          ...transaction,
+          from: this.#address,
+        },
+      },
+    };
+
+    try {
+      const response = await this.executePrivyRequest<PrivyResponse<{ hash: Hex }>>(body);
+      return response.data?.hash;
+    } catch (error) {
+      if (error instanceof Error) {
+        throw new Error(`Transaction sending failed: ${error.message}`);
+      }
+      throw new Error("Transaction sending failed");
+    }
+  }
+
+  /**
+   * Waits for a transaction receipt.
+   *
+   * @param txHash - The hash of the transaction to wait for.
+   * @returns The transaction receipt.
+   */
+  async waitForTransactionReceipt(txHash: Hex): Promise<TransactionReceipt> {
+    return await this.#publicClient.waitForTransactionReceipt({
+      hash: txHash,
+    });
+  }
+
+  /**
+   * Reads data from a smart contract.
+   *
+   * @param params - Parameters for reading the contract
+   * @param params.address - The address of the contract
+   * @param params.abi - The ABI of the contract
+   * @param params.functionName - The name of the function to call
+   * @param params.args - The arguments to pass to the function
+   * @returns A Address that resolves to the contract function's return value
+   */
+  async readContract<
+    const abi extends Abi | readonly unknown[],
+    functionName extends ContractFunctionName<abi, "pure" | "view">,
+    const args extends ContractFunctionArgs<abi, "pure" | "view", functionName>,
+  >(
+    params: ReadContractParameters<abi, functionName, args>,
+  ): Promise<ReadContractReturnType<abi, functionName, args>> {
+    const body = {
+      address: this.#address,
+      chain_type: "ethereum",
+      chain_id: NETWORK_ID_TO_CHAIN_ID[this.#network.chainId!],
+      contract: {
+        address: params.address,
+        abi: params.abi,
+        functionName: params.functionName,
+        args: params.args,
+      },
+    };
+
+    const response = await this.executePrivyRequest<{
+      result: ReadContractReturnType<abi, functionName, args>;
+    }>({
+      method: "eth_call",
+      params: body,
+    });
+    return response.result;
+  }
+
+  /**
+   * Transfer the native asset of the network.
+   *
+   * @param to - The destination address.
+   * @param value - The amount to transfer in Wei.
+   * @returns The transaction hash.
+   */
+  async nativeTransfer(to: string, value: string): Promise<Hex> {
+    const valueInWei = parseEther(value);
+    const valueHex = `0x${valueInWei.toString(16)}`;
+
+    const body = {
+      address: this.#address,
+      chain_type: "ethereum",
+      method: "eth_sendTransaction",
+      caip2: `eip155:${NETWORK_ID_TO_CHAIN_ID[this.#network.chainId!]}`,
```

Comment:
This does not work, as chainId will not be in the network id to chain id lookup, and thus this returns undefined, and `eip155:undefined` ends up in the header.

Please fix by using the chainId outright

```ts
caip2: `eip155:${this.#network.chainId!}`,
```